Propagate group creation properties to intermediate groups

vchoi-hdfgroup opened the pull request:
Propagate object creation properties to intermediate groups.

Fixes #3945

vchoi-hdfgroup committed:
Fix for github issue #3945:
Propagate object creation properties to intermediate groups.

github-actions[bot] committed:
Committing clang-format changes

derobins commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    hid_t    sid         = H5I_INVALID_HID;
+    hsize_t  dims[1]     = {10};
+    unsigned cr_order    = 0;
+    hbool_t  track_times = false;
```
Use bool instead of hbool_t

qkoziol commented on src/H5Gtraverse.c:
```diff
@@ -666,8 +669,14 @@ H5G__traverse_real(const H5G_loc_t *_loc, const char *name, unsigned target, H5G
+                    if (ocrt_info->obj_type == H5O_TYPE_GROUP)
+                        gcrt_info.gcpl_id = ((H5G_obj_create_t *)ocrt_info->crt_info)->gcpl_id;
+                    else if (ocrt_info->obj_type == H5O_TYPE_DATASET)
+                        gcrt_info.gcpl_id = ((H5D_obj_create_t *)ocrt_info->crt_info)->dcpl_id;
```
This seems like a poor reason to include the H5Dpkg header.  Please add a private 'getter' routine to the H5D interface instead.

qkoziol replied:
Actually, it looks like there should be 3 new private routines: a 'getter' in the H5L interface (so the H5L_trav_cr_t struct stays hidden from other packages), and a 'getter' for the H5G and H5D interfaces that retrieve the GCPL & DCPL IDs, respectively.

vchoi-hdfgroup replied:
I implemented the 3 getters for H5L, H5G, and H5D interfaces.
However, I got the following CI build errors: 
 /usr/bin/ld: ../src/.libs/libhdf5.so: undefined reference to `H5D_OBJ_ID'
collect2: error: ld returned 1 exit status

I couldn't figure out what went wrong for the macro H5D_OBJ_ID in the H5D interface.  Anyone knows what might be the problem?

qkoziol replied:
Not certain about this, since H5Gtraverse included H5Dprivate.h. Maybe delete your build directory and re-run autogen.sh?

vchoi-hdfgroup replied:
The failures are from the CI actions. Not from my own build on jelly.

qkoziol reviewed (changes requested):
Adding 'getter' routines will avoid coupling these packages together.

vchoi-hdfgroup committed:
Address CI failures and review feedback.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup committed:
Try to resolve the CI build failures.

vchoi-hdfgroup committed:
Merge branch 'HDFGroup:develop' into issue_3945_develop

vchoi-hdfgroup committed:
Fix the CI build failures for undefined reference to `H5D_OBJ_ID'
which was a macro defined in src/H5Dprivate.h.
The header file "H5Dprivate.h" was removed from src/H5Gtraverse.c
in the commit "Remove useless headers (#4145)".

github-actions[bot] committed:
Committing clang-format changes

qkoziol reviewed (approved):
Nice!

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+     */
+
+    /* Create groups that use default creation properties */
+    def_gid = H5Gcreate2(fid, "def_group1/def_group2", lcpl, H5P_DEFAULT, H5P_DEFAULT);
```
Is it better to use constants for these names?

jhendersonHDF commented on src/H5Gtraverse.c:
```diff
@@ -665,8 +667,15 @@ H5G__traverse_real(const H5G_loc_t *_loc, const char *name, unsigned target, H5G
+                    if (ocrt_info->obj_type == H5O_TYPE_GROUP)
+                        gcrt_info.gcpl_id = H5G_OBJ_ID(ocrt_info->crt_info);
+                    else if (ocrt_info->obj_type == H5O_TYPE_DATASET)
+                        gcrt_info.gcpl_id = H5D_OBJ_ID(ocrt_info->crt_info);
```
Seems like this may be problematic to assign a DCPL ID to a GCPL ID. They're both subclasses of H5P_CLS_OBJECT_CREATE_g, but if the code ever needs to retrieve some group-specific properties (maybe in the future), it seems like that may fail. Should `gcrt_info` have an `ocpl_id` field rather than a `gcpl_id` field to try to make it clear that only generic OCPL properties should be retrieved from it?

qkoziol replied:
> Seems like this may be problematic to assign a DCPL ID to a GCPL ID. They're both subclasses of H5P_CLS_OBJECT_CREATE_g, but if the code ever needs to retrieve some group-specific properties (maybe in the future), it seems like that may fail. Should `gcrt_info` have an `ocpl_id` field rather than a `gcpl_id` field to try to make it clear that only generic OCPL properties should be retrieved from it?

Agree

vchoi-hdfgroup replied:
I have couple questions about changing gcpl_id to ocpl_id:
--Changing gcpl_id field in gcrt_info to ocpl_id would lead to changes in quite a few files which uses gcrt_info.gpcl_id.
 --If I change gcpl_id in gcrt_info which is H5G_obj_create_t struct, should I also change dcpl_id field in H5D_obj_create_t struct to ocpl_id?

jhendersonHDF replied:
> --Changing gcpl_id field in gcrt_info to ocpl_id would lead to changes in quite a few files which uses gcrt_info.gpcl_id.

It's probably fine to put this off to a separate focused PR since the code appears to currently only retrieve generic properties.

> --If I change gcpl_id in gcrt_info which is H5G_obj_create_t struct, should I also change dcpl_id field in H5D_obj_create_t struct to ocpl_id?

As far as I can tell, it looks like the only place that sets up a `H5D_obj_create_t` struct sets its `dcpl_id` field to a DCPL, so that one shouldn't need to be changed.

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    CHECK(status, FAIL, "H5Gclose");
+    VERIFY(track_times, true, "H5Pget_obj_track_times");
+
+    /* Default is false */
```
Default is 0?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    CHECK(status, FAIL, "H5Gclose");
+
+    status = H5Pclose(def_gcpl);
+    CHECK(status, FAIL, "H5Gclose");
```
H5Pclose?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+     * group with intermediate groups
+     */
+    gcpl = H5Pcreate(H5P_GROUP_CREATE);
+    CHECK(gcpl, FAIL, "H5Fcreate");
```
H5Pcreate?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    VERIFY(track_times, false, "H5Pget_obj_track_times");
+
+    status = H5Gclose(gid);
+    CHECK(status, FAIL, "H5Pclose");
```
H5Gclose?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+     * dataset with intermediate group
+     */
+    dcpl = H5Pcreate(H5P_DATASET_CREATE);
+    CHECK(dcpl, FAIL, "H5Fcreate");
```
H5Pcreate?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+
+    /* Verify dname */
+    did = H5Dopen2(gid, "dname", H5P_DEFAULT);
+    CHECK(did, FAIL, "H5Gopen2");
```
H5Dopen2?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    VERIFY(track_times, true, "H5Pget_obj_track_times");
+
+    status = H5Dclose(did);
+    CHECK(status, FAIL, "H5Pclose");
```
H5Dclose?

bmribler commented on test/tmisc.c:
```diff
@@ -6640,6 +6641,223 @@ test_misc39(void)
+    CHECK(status, FAIL, "H5Pclose");
+
+    status = H5Gclose(gid);
+    CHECK(status, FAIL, "H5Pclose");
```
H5Gclose

bmribler reviewed (commented):
Minor typos

vchoi-hdfgroup committed:
Fixes basd on review feedback.
